build/SharedFx.targets

```diff
@@ -121,6 +127,137 @@
+      <CrossGenTool Condition="$([MSBuild]::IsOSPlatform('Windows'))">$(CrossGenTool).exe</CrossGenTool>
+
+      <LibPrefix Condition="$([MSBuild]::IsOSPlatform('Linux')) OR $([MSBuild]::IsOSPlatform('OSX'))">lib</LibPrefix>
+      <LibExtension>.so</LibExtension>
+      <LibExtension Condition="$([MSBuild]::IsOSPlatform('Windows'))">.dll</LibExtension>
+      <LibExtension Condition="$([MSBuild]::IsOSPlatform('OSX'))">.dylib</LibExtension>
+    </PropertyGroup>
+
+    <!-- Determine runtime location -->
+    <MSBuild Projects="$(SharedFxWorkDirectory)SharedFx.csproj" Targets="GetPackageDefinitions" >
+      <Output TaskParameter="TargetOutputs" ItemName="PackageDefinitions" />
+    </MSBuild>
+
+    <ItemGroup>
+      <RuntimePackage Include="@(PackageDefinitions)" Condition="$([System.String]::new('%(PackageDefinitions.Name)').Contains('runtime')) AND $([System.String]::new('%(PackageDefinitions.Name)').Contains('$(RuntimePackageName)'))" />
+      <RuntimePackageFiles Include="%(RuntimePackage.ResolvedPath)\runtimes\**\*" />
+      <CrossGenToolFile Include="%(RuntimePackage.ResolvedPath)\**\$(CrossGenTool)" />
+    </ItemGroup>
+
+    <!-- Create tool directory with crossgen executable and runtime assemblies -->
+    <Copy SourceFiles="@(RuntimePackageFiles);@(CrossGenToolFile)" DestinationFolder="$(SharedFxCrossGenToolDirectory)"/>
+
+    <ItemGroup>
+      <ClrJitAssembly Include="$(SharedFxCrossGenToolDirectory)\**\$(LibPrefix)clrjit$(LibExtension)" />
+    </ItemGroup>
+
+    <Error Text="Expected to resolve a single runtime package but instead resolved @(RuntimePackage->Count()) with identities %(RuntimePackage.Identity)" Condition="'@(RuntimePackage->Count())' != 1" />
+    <Error Text="Expected to resolve a single clr jit assembly but instead resolved @(ClrJitAssembly->Count()) with identities %(ClrJitAssembly.Identity)" Condition="'@(ClrJitAssembly->Count())' != 1" />
+
+    <!-- Gather details on published assemblies -->
+    <MSBuild Projects="$(SharedFxWorkDirectory)SharedFx.csproj"
+      Targets="GetPublishAssemblies"
+      Properties="RuntimeIdentifier=$(SharedFxRID);SelfContained=false" >
+      <Output TaskParameter="TargetOutputs" ItemName="PublishAssemblies" />
+    </MSBuild>
+
+    <ItemGroup>
+      <IgnoredAssemblies Include="@(PublishAssemblies)" Condition="'%(AssetType)' == 'native' OR '%(AssetType)' == 'resources'" />
+      <_AssembliesToCrossgen Include="$(SharedFxOutputDirectory)**\*.dll" />
+    </ItemGroup>
+
+    <!-- Compute the intersection of crossgen candidates and native/resources assemblies as the set of assemblies to skip crossgen -->
+    <CreateItem Include="@(_AssembliesToCrossgen)" Condition="'%(Filename)' != ''and '@(IgnoredAssemblies)' != ''">
+      <Output TaskParameter="Include" ItemName="AssembliesToRemove"/>
+    </CreateItem>
+
+    <!-- Resolve list of assemblies to crossgen -->
+    <ItemGroup>
+      <AssembliesToCrossgen Include="@(_AssembliesToCrossgen)">
```

Comment:
Could be condensed to 
```xml
<AssembliesToCrossgen Include="@(_AssembliesToCrossgen)" Exclude="@(AssembliesToRemove)">
```

Reply:
I tried that, for some reason this did not work and I had to use Remove. I'll test it out a little more to see why not.

Reply:
MSBuild globbing matches can be finicky. For example if you include "C:\myfile.txt" and exclude "C:\dir\..\myfile.txt" it doesn't work, even though the normalized file path is the same